### PR TITLE
add accidentals, slurs and dynamics to data generation

### PR DESCRIPTION
**Pull request by Juphex**
this is required for https://github.com/UHHRobotics22-23/marimbabot/blob/main/marimbabot_vision/scripts/real_world_data_collection.py

add slurs
add repeats
add major minor scales
add articulations
add chords (2 notes)
add dynamics

### CODE REVIEW

**Comment by Juphex** [comment]
TODO add \repeat volta 1 { c1 }
TODO add slurs

**Comment by Juphex** [comment]
https://github.com/UHHRobotics22-23/marimbabot/issues/70#issuecomment-1534731239

**Comment by Juphex** [comment]
I added more music sheet notations. For the chords I stick to 2 arbitrary notes @scaredycode. I guess that is not important whether they make sense. I put some examples below.
I had to adjust the ground truth sequence so it includes the new stuff.

Ground Truth txts for first[1] and second[2] picture:
[1] \key e \minor r4 e''8 - \portato <c'' a''>8 gff'2 eff'1 a'1
[2] repeat volta 2 \key d \minor a''8 \ppp a''8 d'8 ( g'8 <d' g'>4 g''4 r8 b'8 <c' g'>4 g'8 ff'8 f''8 - \accent g''8 d'1 )
Lilypond:
[1]
\context Staff = "Staff_1"
{
    \context Voice = "Voice_1"
    {
        \key e \minor
        r4
        e''8
        - \portato
        <c'' a''>8
        gff'2
        eff'1
        a'1
    }
}`
[2]
\context Staff = "Staff_1"
{
    \repeat volta 2
    \context Voice = "Voice_1"
    {
        \key d \minor
        a''8
        \ppp
        a''8
        d'8
        (
        g'8
        <d' g'>4
        g''4
        r8
        b'8
        <c' g'>4
        g'8
        ff'8
        f''8
        - \accent
        g''8
        d'1
        )
    }
}`
staff_1
staff_1

**Comment by Juphex** [comment]
The data augmentation that we have works fine but, as expected, the handwritten creation does not work @scaredycode . But I guess @Flova  could train the model without it.

**Comment by scaredycode** [comment]
> The data augmentation that we have works fine but, as expected, the handwritten creation does not work @scaredycode . But I guess @Flova  could train the model without it.

You mean the augmentation does not work with generated handwritten data? Or the handwritten data generation does not work at all? I'll look into it.

**Comment by scaredycode** [comment]
Looks really good. Just keep in mind the marimbabot won't be able to play much of the added stuff (accents, legato,...) which is not exactly an issue, just more effort for the vision to correctly recognize, and for other modules to filter out all the stuff not needed for playing (planning, audio feedback,...)
Also the double flats in the first picture decrease the note by two half steps instead of one which the goal position planning will need to keep in mind. To my knowledge this is also rarely used in music. My suggestion would be to start with only the basic accidentals (sharp, flat, natural). 
But of course we can also change that later depending on how well everything works with the added stuff 👍

**Comment by Juphex** [comment]
@scaredycode  handwritten data generation will not work at all because it does not include handwritten data for slurs, dots etc.  when we include all the new stuff while generating data. 


You are right that the marimbabot cannot play everything. I am looking forward on getting input concerning what I should remove here. (e.g. what kind of articulations)

**Comment by Juphex** [comment]
TODO fix: "\repeat" is represented as "repeat" in .txt

**Comment by Juphex** [comment]
as @scaredycode and I discussed this PR does not cover swing (proposed in the all my ducklings example here https://github.com/UHHRobotics22-23/marimbabot/pull/34#issuecomment-1541878041)

**Comment by Juphex** [comment]
``I tested the changes regarding the missing "\" of "\repeat" in the txt file and I did not notice any problems for data samples that do not start with a repeat.
E.g. :

staff_1

txt:
`\tempo 4=40 cf'2 \mf g''4. <e'' a''>8 a'8 e'4 c'4 - \staccatissimo eff'2 r8. css'8 d'4. - \staccatissimo <fff' a'>4 d'4
`
.ly:
```
%! abjad.LilyPondFile._get_format_pieces()
\version "2.20.0"
%! abjad.LilyPondFile._get_format_pieces()
\language "english"

%! abjad.LilyPondFile._get_formatted_blocks()
\header
%! abjad.LilyPondFile._get_formatted_blocks()
{
    tagline = #ff
%! abjad.LilyPondFile._get_formatted_blocks()
}
#(set-default-paper-size "a8" 'landscape)
#(set-global-staff-size 16)
\context Staff = "Staff_1"
{
    \context Voice = "Voice_1"
    {
        \tempo 4=40
        cf'2
        \mf
        g''4.
        <e'' a''>8
        a'8
        e'4
        c'4
        - \staccatissimo
        eff'2
        r8.
        css'8
        d'4.
        - \staccatissimo
        <fff' a'>4
        d'4
    }
}
```

File: marimbabot_vision/scripts/generate_data.py [hunk 1]
```diff
@@ -222,6 +221,8 @@ def generate_sample(i, args):
                                                 include_repeat=args.repeats, include_slurs=args.slurs, min_duration=args.min_duration)
     print(args.output_dir)
 
+    generate_sample(0, args)
```
Comment (Juphex) [inline marimbabot_vision/scripts/generate_data.py:224]: TODO delete @Juphex

**Comment by Juphex** [comment]
@ateRstones Thank you for your inquiry. The current state of this pull request is that it is awaiting review. Although there hasn't been any recent activity in the past three weeks, I am eagerly waiting for a reviewer to provide feedback and give their approval for merging. If you have any further questions or concerns, please feel free to let me know. Thank you!

**Comment by Juphex** [comment]
This pull request needs really a review. 😁

@[ateRstones](https://github.com/ateRstones)
ateRstones
@[Flova](https://github.com/Flova)
Flova
@[iibrahimli](https://github.com/iibrahimli)
iibrahimli
@[berkgungor](https://github.com/berkgungor)
berkgungor
@[scaredycode](https://github.com/scaredycode)
scaredycode
@[yunlong-wang-cn](https://github.com/yunlong-wang-cn)
yunlong-wang-cn
@[cvhex](https://github.com/cvhex)
cvhex

**Review by scaredycode — APPROVED** [review]
I'm still not a very big fan of the double sharps/flats, but since the whiteboard sample collection is already in progress with this level of detail, we might as well keep it now. 
Also there is still this small issue with the broken bars due to dotted notes increasing the beats per bar (exceeding 4/4), but these notes should still be playable (although very very illegal from a musician pov xD) 
Rest looks fine 👍

**Review by Flova — APPROVED** [review]
I am also not a fan of the sometimes in valid configurations, but it otherwise looks good imo.

**Pull request merged**